### PR TITLE
Cocoa Metal example (no SDL).

### PR DESCRIPTION
**Pull request by yay**
- [x] Have you added the example to the CI at `.github/workflows/check.yml`?

Screenshot 2024-08-20 at 11 50 53 PM

### CODE REVIEW

**Comment by yay** [comment]
PR to fix build from failing (vet tabs): https://github.com/odin-lang/Odin/pull/4117

File: ols.json
Comment (laytan) [inline ols.json:1]: Can you remove this file? You may also add it to .gitignore

**Comment by mbger** [comment]
Just took the sample for a spin and here are a few observations. Hope you don't mind me chiming in.

The other samples all have a way of exiting the main loop (e.g. pressing escape to set a `quit` flag to `true`). This is missing here. Therefor the compiler does not complain that there's no return statement at the end of `metal_main`. When I implemented this 'press escape to quit' functionality the program would segfault when exiting `metal_main`. I *think* this happens because `metal_main` sets up a scoped autorelease pool? I.e. when exiting `metal_main` already-released objects get released again wich causes the segfault.

Also, the program currently fetches just one event per main-loop iteration instead of draining all events from the queue. But this might be intentional?

Those were essentially my modifications:

```odin
// `metal_main`
{
    // Setup... (but remove call to NS.scoped_autoreleasepool() to prevent segfault)

    // Main Loop
    for quit := false; !quit;
    {
	// Poll all events from the queue
	for
	{
	    e := app->nextEventMatchingMask(NS.EventMaskAny, nil, NS.DefaultRunLoopMode, true)
	    
	    if e == nil { break }
	    
	    #partial switch e->type()
	    {
		case .KeyDown, .KeyUp:
		code := NS.kVK(e->keyCode())
		#partial switch code {
		    case .Escape:
		    quit = true
		    case .LeftArrow:
		    angle -= 0.02
		    case .RightArrow:
		    angle += 0.02
		    case:
		    fmt.println(code)
		}
		
		case:
		fmt.println(e->locationInWindow(), e->modifierFlags())
		app->sendEvent(e)
	    }
	}

	// Render frame...
    }

    // Exit `metal_main`
    return nil
}
```

File: ols.json
Comment (yay) [reply ols.json:1]: Ah, didn't mean to commit it.

**Comment by yay** [comment]
@mbger Thanks! I added the Escape key handling. The reason I used `NS.Date.distantFuture()` is to keep CPU usage at 0% on idle. If use poll all events and use `nil` instead, idle CPU usage is quite high.

I wonder if there's a memory leak somewhere. If I press left or right key and hold it for a minute, memory usage for the process will slowly keep increasing. With or without the `NS.scoped_autoreleasepool()` statement.

**Comment by mbger** [comment]
> I wonder if there's a memory leak somewhere? If I press left or right key and hold it for a minute, memory usage for the process will slowly keep increasing. With or without the `NS.scoped_autoreleasepool()` statement.

In your original example, the scoped autorelease pool never has a chance to release anything because the relevant scope never exits. I think what might help is to set up a scoped autorelease pool specifically for the event-polling section of our code (though I can't test now): 

```odin
    // Main Loop
    for quit := false; !quit;
    {
	// Poll all events from the queue
	{   
	    NS.scoped_autoreleasepool()
	    
	    for
	    {
		e := app->nextEventMatchingMask(NS.EventMaskAny, nil, NS.DefaultRunLoopMode, true)
		
		if e == nil { break }
		
		#partial switch e->type()
		{
		    case .KeyDown, .KeyUp:
		    code := NS.kVK(e->keyCode())
		    #partial switch code {
			case .Escape:
			quit = true
			case .LeftArrow:
			angle -= 0.02
			case .RightArrow:
			angle += 0.02
			case:
			fmt.println(code)
		    }
		    
		    case:
		    fmt.println(e->locationInWindow(), e->modifierFlags())
		}
		app->sendEvent(e)
	    }
	}

	// Render frame
```

**Comment by yay** [comment]
> I think what might help is to set up a scoped autorelease pool specifically for the event-polling section of our code

Yes, that's definitely an improvement. Just made that change. It still looks like it might be leaking a little though.

**Comment by laytan** [comment]
One thing I've noticed is that pressing the x to close the app just closes the window but doesn't quit the program.

**Comment by yay** [comment]
Yes, one needs to create an AppDelegate and implement one of its methods to handle x button behavior. But IDK how to combine that and explicit event loop handling.

**Comment by laytan** [comment]
> Yes, one needs to create an AppDelegate and implement one of its methods to handle x button behavior. But IDK how to combine that and explicit event loop handling.

That seems like a blocker to me, any application should support properly closing like that

**Comment by yay** [comment]
No problem. Should I keep this open or close and reopen when I have an idea how to do that?

I don't have any immediate plans to explore further, the AppDelegate stuff is way over my head and I'm making more progress elsewhere.

**Comment by laytan** [comment]
You don't have to close it, maybe somebody else sees it and is able to implement the missing things.

**Comment by colrdavidson** [comment]
I don't know how useful it is, but I did something like this a while back
https://gist.github.com/colrdavidson/aecf689196fcae49122749f866780854#file-osx_platform-odin-L190

**Comment by mbger** [comment]
> I don't have any immediate plans to explore further, the AppDelegate stuff is way over my head and I'm making more progress elsewhere.

I think you're in good company. This stuff is tricky to get right and after doing a bit of research I think every approach will have its drawbacks. See for example [this discussion](https://github.com/floooh/sokol/pull/483) in the sokol project. (AFAICT they don't set up their own main loop but start the `NSApp` run loop and hook int the main window's `drawRect:` to render each frame)

As for the immediate PR-blocking issue here (exiting the program when the window is closed): I think we could register a [NSWindowDelegate](https://developer.apple.com/documentation/appkit/nswindowdelegate) and set our `quit` flag in `windowWillClose:`

**Comment by yay** [comment]
@mbger Thank you! Just pushed those changes. Now there are 4 different ways to close the app. From the menu, via Cmd+Q shortcut, by pressing Escape, and by clicking the x button.

@colrdavidson Thanks! But I don't see where the `get_next_event` proc is being called?

**Comment by mbger** [comment]
Don't know if that's the way to do it, but what worked for me is to create a custom subclass of `NSWindow` which overrides `keyDown:`. Not sure if that approach breaks some other stuff down the road though... (I don't have a ton of experience with macOS dev)

```odin
USE_CUSTOM_WINDOW :: true

    // ...

    window: ^NS.Window

    when USE_CUSTOM_WINDOW
    {	
	customWindowClass := NS.objc_allocateClassPair(intrinsics.objc_find_class("NSWindow"), strings.clone_to_cstring("CustomWindow", context.temp_allocator), 0)

	keyDown :: proc "c" (event: ^NS.Event) { /* ignore key down events */ }

	assert(customWindowClass != nil)
	
	NS.class_addMethod(customWindowClass, intrinsics.objc_find_selector("keyDown:"), auto_cast keyDown, "v@:@")
	NS.objc_registerClassPair(customWindowClass)
	keyIgnoringWindow := NS.class_createInstance(customWindowClass, size_of(NS.Window))
	window = cast(^NS.Window)keyIgnoringWindow
    }
    else
    {
	window = NS.Window.alloc()
    }

    defer window->release()
```

**Comment by yay** [comment]
Apparently, one can override the `acceptsFirstResponder` property to return `true`: https://stackoverflow.com/questions/13519647/silence-cocoa-error-beep

But I haven't figured out how to do that yet. Is that the same as overriding a method?

There are these ObjC Runtime functions:
```
class_addProperty           :: proc(cls: Class, name: cstring, attributes: [^]objc_property_attribute_t, attributeCount: uint) -> BOOL ---
	class_replaceProperty       :: proc(cls: Class, name: cstring, attributes: [^]objc_property_attribute_t, attributeCount: uint) ---
```
But I'm reading that you can't actually add properties at runtime: https://stackoverflow.com/questions/7819092/how-can-i-add-properties-to-an-object-at-runtime

So I'm confused.

In the meantime, I've submitted this as I keep researching: https://github.com/odin-lang/Odin/pull/4171

**Comment by mbger** [comment]
> Apparently, one can override the `acceptsFirstResponder` property to return `true`: https://stackoverflow.com/questions/13519647/silence-cocoa-error-beep
>
> 
> [...] 
> So I'm confused. This doesn't make it stop beeping on key press:
> 

Maybe the accepted answer is just incorrect and there's nothing wrong with how your Odin code? Have you tried to confirm the accepted answer with Xcode > File > New Project?

**Comment by yay** [comment]
> Maybe the accepted answer is just incorrect and there's nothing wrong with how your Odin code? Have you tried to confirm the accepted answer with Xcode > File > New Project?

Well, `acceptsFirstResponder` indeed doesn't work in both Objective-C and Odin.
But `performKeyEquivalent` does! And key presses are silent.

```
performKeyEquivalent :: proc "c" (event: ^NS.Event) -> NS.BOOL { return NS.YES }
NS.class_addMethod(CustomWindowClass, intrinsics.objc_find_selector("performKeyEquivalent:"), auto_cast performKeyEquivalent, "B@:@")
```

However, Cmd+Q stops working.

**Comment by mbger** [comment]
Just curious: What are the downsides with the approach I outlined in a [previous comment](https://github.com/odin-lang/examples/pull/52#issuecomment-2315583106)? At least on my end everything seems to work just fine – no beeps and CMD+Q quits the program.

**Comment by yay** [comment]
I would prefer not to ignore the key down event, if possible. My intent was to show how to create a macOS GUI that idles with zero CPU usage in the absence of events. And when typing, characters show up on screen on key down event. Though it's probably equally as important for games.

Of course, I can just use your solution (which I'm grateful for btw) just to get this PR to merge. But I wonder if there's something else we can try? This has been a very productive conversation so far.

**Comment by yay** [comment]
@mbger Actually, never mind my previous comment. I misunderstood the impact of overriding the `keyDown` method. We still do receive the key down event in the event loop even then and get a chance to process it. Thanks for your help!

@laytan I think this is ready to merge now.

**Comment by laytan** [comment]
2 things I still see that I wouldn't expect:

1. The window does not start in focus:

https://github.com/user-attachments/assets/948c6e50-17e3-4819-a643-92f7bbee80f5

2. Resizing lags behind / does weird things:

https://github.com/user-attachments/assets/430d9aa6-57e4-43fb-b979-36b326de5cdf

**Comment by mbger** [comment]
> 2 things I still see that I wouldn't expect:
> 
> 1. The window does not start in focus:
> 
>  no-focus.mov 
> 
> 2. Resizing lags behind / does weird things:
> 
>  resize-lag.mov

Hm I don't experience those issues with my code. @yay FWIW here's my code: https://gist.github.com/mbger/da25aceed721ec91930a32c4447883f0

**Comment by yay** [comment]
Thanks. I did experience the same issues:

1. Was seemingly fixed by processing all events in the queue on every loop iteration.
2. Was fixed by calling `app->activate()`.

**Comment by laytan** [comment]
I am still not getting the app to start in the foreground, the resizing stuff is fixed though.

**Comment by laytan** [comment]
Looking at Sokol and GLFW, they do `[NSApp activateIgnoringOtherApps:YES];` which when applied here also works but it is deprecated.

**Comment by mbger** [comment]
Instead of 

```odin
app->activate()
```
try
```odin
window->makeKeyAndOrderFront(nil)
intrinsics.objc_send(nil, app, "finishLaunching")
```

[`finishLaunching`](https://developer.apple.com/documentation/appkit/nsapplication/1428771-finishlaunching?language=objc#) is usually called with `[NSApp run]` which we don't call, since we're rolling our own loop.

**Comment by laytan** [comment]
That doesn't work either.

**Comment by mbger** [comment]
That's weird. I copy pasted the latest commit from this PR and made the mentioned change and it works. Would you mind running the code from the Gist I posted earlier? Would love to know if that also doesn't work for you. https://gist.github.com/mbger/da25aceed721ec91930a32c4447883f0

**Comment by laytan** [comment]
Yep, your gist also doesn't work:

https://github.com/user-attachments/assets/56df53ec-a1a6-46b0-b033-1af82a1fed31

**Comment by mbger** [comment]
Thanks! Are you running your browser & terminal/editor in some special windowing mode? Looks a bit like the macOS native fullscreen split mode but w/o the divider in the middle. For me it looks like this: 


https://github.com/user-attachments/assets/fe38c99a-4bc3-49c1-b8bd-3a86d41d3e74

**Comment by laytan** [comment]
> Thanks! Are you running your browser & terminal/editor in some special windowing mode? Looks a bit like the macOS native fullscreen split mode but w/o the divider in the middle. For me it looks like this:

Nothing special going on afaict, I do have Yabai but disabling that doesn't help, my terminal is Alacritty.
It all works as expected with the SDL version and any other app I have tried.

I think we should just put in the deprecated `app->activateIgnoringOtherApps(true)` since that is what GLFW and Sokol do (I haven't looked at what SDL does).

**Comment by yay** [comment]
Weird,

```
window->makeKeyAndOrderFront(nil)
app->finishLaunching()
```

is what I tried and it didn't work for me.

But `activate` did the trick.

I'm on macOS 15.0 Beta (24A5331b), if that matters.

**Comment by yay** [comment]
@laytan I replaced `app->activate()` with `app->activateIgnoringOtherApps(true)`.

**Comment by yay** [comment]
Hmm... Screenshot 2024-09-04 at 11 40 21 PM

So what do we do to merge this PR?

**Comment by laytan** [comment]
https://github.com/odin-lang/Odin/commit/ddf5ca7adf16b87c202989537840d6099c553638 👿

**Comment by yay** [comment]
> [odin-lang/Odin@ddf5ca7](https://github.com/odin-lang/Odin/commit/ddf5ca7adf16b87c202989537840d6099c553638) 👿

Cool. I think we're good to go now.

**Comment by yay** [comment]
This is amazing actually. This has been annoying me for years! On any platform.

> In macOS Sonoma, we have introduced Cooperative App Activation. Cooperative app activation reduces unexpected application switches, for example, an app switch while you are in the middle of typing.
> 
> There are two parts to cooperative app activation. Activate is now a request, as opposed to a command. The system considers the broader context of what the user is doing to decide if the activate request is appropriate. The new yield API allows an application to influence the context of a future activation request.
> 
> Now that activate is a request, the ignoringOtherApps parameter and option are ignored. As such, in macOS Sonoma, the activate(ignoringOtherApps:) function and the activateIgnoringOtherApps option are both deprecated. Replace them with the new activate APIs for NSApplication and NSRunningApplication.

More here: https://developer.apple.com/videos/play/wwdc2023/10054/

So I think we should change back to `app->activate()`.

**Comment by laytan** [comment]
Well as we have seen here, it clearly doesn't work well enough yet. I am guessing apps that open others have to implement something (that yield concept) and until they do just an ->activate doesn't work.

**Comment by yay** [comment]
Not sure how `activateIgnoringOtherApps` would work though, if `Now that activate is a request, the ignoringOtherApps parameter and option are ignored.`

Either way, I'm happy for this to be merged as is too.

**Comment by laytan** [comment]
> Now that activate is a request, the ignoringOtherApps parameter and option are ignored.

It looks like this is not yet the case. I think it's good to leave like this so it behaves like glfw, SDL, and Sokol does, we can always revisit later.

I will look over the entire pr once more in a bit and merge if there is nothing else, great work here any way!

**Comment by laytan** [comment]
Ah man it is not quite there yet, sorry :(

If you minimise the app you can't get it back into focus by clicking it in the dock.

You also can't quit the app by right clicking the dock entry and clicking on quit.

**Comment by mbger** [comment]
> Ah man it is not quite there yet, sorry :(
> 
> If you minimise the app you can't get it back into focus by clicking it in the dock.
> 
> You also can't quit the app by right clicking the dock entry and clicking on quit.

Weird that we're experiencing such different behavior. 

If I minimize it's still active and receives events. And right-click -> quit from the dock icon quits the program.

**Comment by laytan** [comment]
> Weird that we're experiencing such different behavior.

Oh that is weird, I assumed this would reproduce for you too. I did confirm that the sdl version worked as expected, but let me investigate some more.

**Comment by laytan** [comment]
That small example and comparison allowed me to debug it and I found the issue, working on a fix.

**Comment by yay** [comment]
Is it something to do with `objc_msgSend`? Or Application not extending NSResponder? I've been trying to figure it out all evening.

**Comment by laytan** [comment]
It was because our `EventMaskAny` was wrong, now fixed through https://github.com/odin-lang/Odin/commit/27ed10746db72fd3e5ba9aea922cb064aa3f8995 and https://github.com/odin-lang/Odin/commit/201a7b90bb52facf9bd364a96061fc6c19e19031

**Comment by laytan** [comment]
I believe the last thing you need here is a call to `app->finishLaunching()`

**Comment by yay** [comment]
Ah, so the right events were not coming through.

> I believe the last thing you need here is a call to `app->finishLaunching()`

Done, thanks. Indeed, the dock menu Quit didn't work without this one even with the new mask.

**Comment by laytan** [comment]
Thanks for sticking with me and all the work here, this is a very nice example to accompany the native win32 example!

**Comment by mbger** [comment]
Nice, thank you!

**Pull request merged**